include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1523 @@
+
+                if (layout.area() < best_optimized_solution)
+                {
+                    best_optimized_solution = layout.area();
+
+                    if (ps.verbose)
+                    {
+                        print_placement_info(layout);
+                    }
+
+                    return {{}, layout};
+                }
+
+                return {{}, std::nullopt};
+            }
+
+            if (improv_mode && area >= best_solution)
+            {
+                return {{}, std::nullopt};
+            }
+
+            // Expand layout dimensions if needed
+            if (position.x == layout.x() && layout.x() < (max_layout_width - 1) &&
+                !ssg.network.is_po(ssg.nodes_to_place[current_node - 1]))
+            {
+                layout.resize({layout.x() + 1, layout.y(), layout.z()});
+            }
+            if (position.y == layout.y() && layout.y() < (max_layout_height - 1) &&
+                !ssg.network.is_po(ssg.nodes_to_place[current_node - 1]))
+            {
+                layout.resize({layout.x(), layout.y() + 1, layout.z()});
+            }
+
+            // Check if it's the last position in the current vertex
+            if (idx == (ssg.current_vertex.size() - 1))
+            {
+                if (!valid_layout(layout, ssg.network, node2pos, ssg.nodes_to_place, current_node))
+                {
+                    return {{}, std::nullopt};
+                }
+
+                possible_positions = get_possible_positions(layout, ssg, current_node, node2pos);
+            }
+        }
+
+        // Generate next possible positions with their priorities
+        for (const auto& position : possible_positions)
+        {
+            auto new_sequence = ssg.current_vertex;
+            new_sequence.push_back(position);
+
+            const auto remaining_nodes_to_place =
+                static_cast<double>(ssg.nodes_to_place.size() - (ssg.current_vertex.size() + 1));
+
+            // Current layout size
+            const double cost1 = static_cast<double>(((std::max(layout.x() - 1, position.x) + 1) *
+                                                      (std::max(layout.y() - 1, position.y) + 1))) /
+                                 static_cast<double>((max_layout_width * max_layout_height));
+
+            // Position of last placed node
+            const double cost2 = static_cast<double>(((position.x + 1) * (position.y + 1))) /
+                                 static_cast<double>((max_layout_width * max_layout_height));
+
+            double priority = remaining_nodes_to_place + cost1 + cost2;
+            next_positions.push_back({new_sequence, priority});
+        }
+
+        return {next_positions, std::nullopt};
+    }
+    /**
+     * Initializes the allowed positions for primary inputs (PIs), the cost for each search space graph and the maximum
+     * number of expansions.
+     */
+    void initialize_pis_cost_and_num_expansions() noexcept
```

Comment:
All these initialize functions are quite wild and host a lot of magic numbers. Can we extract constants from them? I.e., `static constexpr uint64_t meaningful_name = 4ul;`